### PR TITLE
refactor(ledger): switch blockstore to wincode

### PR DESCRIPTION
**Pull request by kskalski**
#### Problem
Blockstore data is serialized with bincode, which can be replaced by wincode serialization to improve performance.

#### Summary of Changes
* add wincode schema annotations / impl
* remove unused serde annotations / impl from `ledger/`

Note, the whole `ledger` crate can almost get rid of `bincode`, except for a small change (outside of blockstore) depending on https://github.com/anza-xyz/solana-sdk/pull/649, so it is left out for follow-up PR.

### CODE REVIEW

File: ledger/src/blockstore/column.rs [hunk 1]
```diff
@@ -291,14 +298,15 @@ pub trait ColumnName {
 
 // Columns that serialize data on insertion and deserialize on fetch
 pub trait TypedColumn: Column {
-    type Type: Serialize + DeserializeOwned;
+    type Type: SchemaWrite<DefaultConfig, Src = Self::Type>
+        + SchemaReadOwned<DefaultConfig, Dst = Self::Type>;
 
     fn deserialize(data: &[u8]) -> Result<Self::Type> {
```
Comment (cpubot) [inline ledger/src/blockstore/column.rs:315]: Should `#[inline]` this

File: ledger/src/blockstore/column.rs [hunk 2]
```diff
@@ -589,7 +597,7 @@ impl TypedColumn for columns::Index {
     type Type = blockstore_meta::Index;
 
     fn deserialize(data: &[u8]) -> Result<Self::Type> {
```
Comment (cpubot) [inline ledger/src/blockstore/column.rs:612]: Should `#[inline]` this

File: ledger/src/bit_vec.rs [hunk 3]
```diff
@@ -39,28 +41,32 @@ impl<const NUM_BITS: usize> Default for BitVec<NUM_BITS> {
...
-
-        Ok(Self {
-            words: words.into_boxed_slice(),
-        })
+unsafe impl<'de, const NUM_BITS: usize, C: Config> SchemaRead<'de, C> for BitVec<NUM_BITS> {
+    type Dst = Self;
+
+    fn read(mut reader: impl Reader<'de>, dst: &mut MaybeUninit<Self::Dst>) -> ReadResult<()> {
+        let bytes_len = C::LengthEncoding::read(reader.by_ref())?;
+        if bytes_len > Self::NUM_WORDS {
```
Comment (cpubot) [inline ledger/src/bit_vec.rs]: This isn't incorrect relative to how wincode typically encodes sequences, but it is a departure from the existing behavior. Current behavior reads all `bytes_len` encoded bytes and truncates extraneous bytes.

File: ledger/src/blockstore_meta.rs [hunk 4]
```diff
@@ -176,42 +180,54 @@ impl From<SlotMetaV3> for SlotMeta {
...
 
-    pub(super) fn serialize<S>(val: &Option<u64>, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        val.unwrap_or(u64::MAX).serialize(serializer)
+    pub(crate) struct OptionCompat;
+    unsafe impl<'de, C: ConfigCore> SchemaRead<'de, C> for OptionCompat {
+        type Dst = Option<Slot>;
+
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs:209]: Could add `TypeMeta` here for `Static` optimization
```rs
const TYPE_META: wincode::TypeMeta =
    <Slot as SchemaRead<'de, C>>::TYPE_META.keep_zero_copy(false);
```

File: ledger/src/blockstore_meta.rs [hunk 5]
```diff
@@ -221,40 +237,45 @@ pub struct ErasureMeta {
...
+        wincode::{
+            ReadResult, WriteResult,
+            config::ConfigCore,
+            io::{Reader, Writer},
+        },
+    };
+
+    /// Serializes a `u32` as `u64` for backward compatibility with on-disk data
+    /// that stored the field with 8-byte width.
+    pub(super) struct U32AsU64;
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs:271]: Is there value in keeping this generic like the `serde` version? e.g,
```rs
/// Serializes a `Src` as `Dst` and deserializes a `Dst` as `Src`.
pub(super) struct Cast<Src, Dst>(PhantomData<(Src, Dst)>);
unsafe impl<'de, C: ConfigCore, Src, Dst> SchemaRead<'de, C> for Cast<Src, Dst>
where
    Dst: SchemaRead<'de, C>,
    Src: TryFrom<Dst::Dst>,
{
    type Dst = Src;

    const TYPE_META: wincode::TypeMeta = Dst::TYPE_META.keep_zero_copy(false);

    #[inline]
    fn read(reader: impl Reader<'de>, dst: &mut MaybeUninit<Self::Dst>) -> ReadResult<()> {
        let val = <Dst as SchemaRead<'de, C>>::get(reader)?;
        let x = Src::try_from(val)
            .map_err(|_| ReadError::Custom("Could not convert dst to src"))?;
        dst.write(x);
        Ok(())
    }
}

unsafe impl<C: ConfigCore, Src, Dst> SchemaWrite<C> for Cast<Src, Dst>
where
    Dst: SchemaWrite<C>,
    Dst::Src: TryFrom<Src>,
    Src: Copy,
{
    type Src = Src;

    const TYPE_META: wincode::TypeMeta = Dst::TYPE_META.keep_zero_copy(false);

    #[inline]
    fn size_of(src: &Self::Src) -> WriteResult<usize> {
        <Dst as SchemaWrite<C>>::size_of(
            &Dst::Src::try_from(*src)
                .map_err(|_| WriteError::Custom("Could not convert src to dst"))?,
        )
    }

    #[inline]
    fn write(writer: impl Writer, src: &Self::Src) -> WriteResult<()> {
        <Dst as SchemaWrite<C>>::write(
            writer,
            &Dst::Src::try_from(*src)
                .map_err(|_| WriteError::Custom("Could not convert src to dst"))?,
        )
    }
}
```

If not used elsewhere, probably fine as is.  Would just add `TypeMeta` for completeness and `Static` optimizations

File: storage-proto/src/lib.rs [hunk 6]
```diff
@@ -17,19 +18,75 @@ use {
...
+
+        fn read(
+            reader: impl Reader<'de>,
+            dst: &mut MaybeUninit<Self::Dst>,
+        ) -> wincode::ReadResult<()> {
+            match <T as SchemaRead<'de, C>>::read(reader, dst) {
+                Ok(()) => Ok(()),
+                Err(ReadError::Io(IoReadError::ReadSizeLimit(_))) => {
+                    dst.write(T::default());
+                    Ok(())
```
Comment (cpubot) [inline storage-proto/src/lib.rs]: Could be `Dst` aware.
```suggestion
        T: SchemaRead<'de, C> + Default,
        T::Dst: Default,
    {
        type Dst = T::Dst;

        fn read(
            reader: impl Reader<'de>,
            dst: &mut MaybeUninit<Self::Dst>,
        ) -> wincode::ReadResult<()> {
            match <T as SchemaRead<'de, C>>::read(reader, dst) {
                Ok(()) => Ok(()),
                Err(ReadError::Io(IoReadError::ReadSizeLimit(_))) => {
                    dst.write(T::Dst::default());
```

File: storage-proto/src/lib.rs [hunk 7]
```diff
@@ -17,19 +18,75 @@ use {
...
+                Err(e) => Err(e),
+            }
+        }
+    }
+
+    unsafe impl<C: Config, T> SchemaWrite<C> for DefaultOnEmptyRead<T>
+    where
+        T: SchemaWrite<C, Src = T>,
+    {
+        type Src = T;
```
Comment (cpubot) [inline storage-proto/src/lib.rs]: Could also be `Dst` aware

```suggestion
        T: SchemaWrite<C>,
    {
        type Src = T::Src;
```

File: storage-proto/src/lib.rs [hunk 8]
```diff
@@ -17,19 +18,75 @@ use {
...
+
```
Comment (cpubot) [inline storage-proto/src/lib.rs:62]: Could use `T`'s `TYPE_META` for serialization

File: storage-proto/src/lib.rs [hunk 9]
```diff
@@ -17,19 +18,75 @@ use {
...
+    /// exhausted (EOF), for backward compatibility when new fields are appended
+    /// to a struct. Equivalent to `#[serde(deserialize_with = "default_on_eof")]`.
+    pub(super) struct DefaultOnEmptyRead<T>(PhantomData<T>);
+
+    unsafe impl<'de, C: Config, T> SchemaRead<'de, C> for DefaultOnEmptyRead<T>
+    where
+        T: SchemaRead<'de, C, Dst = T> + Default,
+    {
+        type Dst = T;
+
```
Comment (cpubot) [inline storage-proto/src/lib.rs:44]: I know you didn't add `TYPE_META` here, but just posting for documentation purposes.

`TypeMeta::Static` should never be applicable here (e.g., derived from a `T` where `T`'s `TYPE_META` is `Static`), as `TypeMeta::Static` indicates that an outer containing type (e.g., a `struct`, `Vec`, etc) is permitted to use `as_trusted_for`, which would perform a bounds-check on the total expected size, which would fail if this particular type wasn't encoded in the serialized payload, thus breaking the expectation of `eof` being accepted.

File: ledger/src/blockstore_meta.rs (hunk 5 above)
Comment (kskalski) [reply ledger/src/blockstore_meta.rs:271]: the serde compat code is removed, so this is the only use-case and I think it doesn't make sense to over-generalize here

I guess we should at some point look at all the custom helpers out there and consider providing util schemas in some wincode sub-module where we can make those generalized - maybe one for casing and other for `From` implementers

File: ledger/src/blockstore_meta.rs (hunk 4 above)
Comment (kskalski) [reply ledger/src/blockstore_meta.rs:209]: sure, added in https://github.com/anza-xyz/agave/pull/11624/commits/f1d0127b94b9fd071c76e0345d4505a79b2a3c26

File: storage-proto/src/lib.rs (hunk 6 above)
Comment (kskalski) [reply storage-proto/src/lib.rs]: applied in [b2bfee5](https://github.com/anza-xyz/agave/pull/11624/commits/b2bfee5f55498850e397f7d8d8e3c35cfadcd582)

File: storage-proto/src/lib.rs (hunk 7 above)
Comment (kskalski) [reply storage-proto/src/lib.rs]: yap, added in [b2bfee5](https://github.com/anza-xyz/agave/pull/11624/commits/b2bfee5f55498850e397f7d8d8e3c35cfadcd582)

File: storage-proto/src/lib.rs (hunk 8 above)
Comment (kskalski) [reply storage-proto/src/lib.rs:62]: yap, added in [b2bfee5](https://github.com/anza-xyz/agave/pull/11624/commits/b2bfee5f55498850e397f7d8d8e3c35cfadcd582)

File: storage-proto/src/lib.rs (hunk 9 above)
Comment (kskalski) [reply storage-proto/src/lib.rs:44]: added comment in [b2bfee5](https://github.com/anza-xyz/agave/pull/11624/commits/b2bfee5f55498850e397f7d8d8e3c35cfadcd582)

File: ledger/src/bit_vec.rs (hunk 3 above)
Comment (kskalski) [reply ledger/src/bit_vec.rs]: Ok, updated in 
https://github.com/anza-xyz/agave/pull/11624/commits/9a37ad24e8260dc6f25e89273eaa21cb82f328a8

We found a use-case for removed 'consume' ;) though `as_trusted_for` can serve as ugly equivalent

File: ledger/src/blockstore/column.rs (hunk 2 above)
Comment (kskalski) [reply ledger/src/blockstore/column.rs:612]: https://github.com/anza-xyz/agave/pull/11624/commits/f1d0127b94b9fd071c76e0345d4505a79b2a3c26

File: ledger/src/bit_vec.rs (hunk 3 above)
Comment (cpubot) [reply ledger/src/bit_vec.rs]: We can't rely on `as_trusted_for` here generally, since the default returns `Ok(self)` it's not guaranteed to actually advance (probably should make this clear in the docs now since that semantic change). Probably best to just use
```rs
let mut vec = <Vec<u8> as SchemaRead<C>>::get(reader.by_ref())?;
vec.resize(Self::NUM_WORDS, 0);
dst.write(Self {
    words: vec.into_boxed_slice()
});
```

Or if we want to prevent a hypothetical over-copy 
```rs
let len = C::LengthEncoding::read(reader.by_ref())?;
let bytes = reader.take_scoped(len)?;
let mut words = Vec::with_capacity(Self::NUM_WORDS);
words.extend_from_slice(&bytes[..bytes.len().min(Self::NUM_WORDS)]);
words.resize(Self::NUM_WORDS, 0u8);
dst.write(Self {
    words: words.into_boxed_slice(),
});
```
Or the equivalent Box uninit route with `take_scoped`

Comment (kskalski) [reply ledger/src/bit_vec.rs]: huh, ok, changed to simple vec deserialize, I think we no longer expect any other len than NUM_WORDS on the wire, so it should be fine to use the resize just for sanitization and do not care about always allocating >= NUM_WORDS (though it could be done with ctx len)

Comment (kskalski) [reply ledger/src/bit_vec.rs]: and this brings up the need for `advance(amt)` API - we could even default implement it using `take_scoped`

**Comment by kskalski** [comment]
Rebased after conflict with https://github.com/anza-xyz/agave/pull/11611 -> I changed `serde_bytes::BytesBuf` into simple `Vec<u8>`, I think bincode doesn't distinguish between these kinds of collections and the APIs used are basically vector's.

**Comment by kskalski** [comment]
Rebased and fixed some typos due to / from https://github.com/anza-xyz/agave/pull/11751

File: ledger/src/blockstore/column.rs (hunk 1 above)
Comment (kskalski) [reply ledger/src/blockstore/column.rs:315]: missed that, 7b2569cf7a

File: ledger/src/blockstore_meta.rs [hunk 10]
```diff
@@ -232,40 +253,49 @@ pub struct ErasureMeta {
...
+    unsafe impl<'de, C: ConfigCore> SchemaRead<'de, C> for U32AsU64 {
+        type Dst = u32;
+
+        const TYPE_META: wincode::TypeMeta = <u64 as SchemaRead<'de, C>>::TYPE_META;
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs]: I think we need `.keep_zero_copy(false)` right?

File: ledger/src/blockstore_meta.rs [hunk 11]
```diff
@@ -232,40 +253,49 @@ pub struct ErasureMeta {
...
-    where
-        D: Deserializer<'de>,
-        R: Deserialize<'de>,
-        T: TryFrom<R>,
-        <T as TryFrom<R>>::Error: std::fmt::Display,
-    {
-        R::deserialize(deserializer)
-            .map(T::try_from)?
-            .map_err(serde::de::Error::custom)
+        const TYPE_META: wincode::TypeMeta = <u64 as SchemaWrite<C>>::TYPE_META;
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs]: Similar to the above, I think we want to prevent zero copy `.keep_zero_copy(false)`

File: ledger/src/blockstore_meta.rs (hunk 10 above)
Comment (kskalski) [reply ledger/src/blockstore_meta.rs]: Yap, fixed

File: ledger/src/blockstore_meta.rs (hunk 11 above)
Comment (kskalski) [reply ledger/src/blockstore_meta.rs]: Right

**Comment by kskalski** [comment]
Also added a few `repr(X)` attributes... I suppose we should have a way for warning that struct could be zero-copy eligible if only it hard proper repr...

File: ledger/src/blockstore_meta.rs [hunk 12]
```diff
@@ -232,40 +253,52 @@ pub struct ErasureMeta {
...
+        }
+
+        fn write(writer: impl Writer, src: &Self::Src) -> WriteResult<()> {
+            <u64 as SchemaWrite<C>>::write(writer, &u64::from(*src))
+        }
     }
 }
 
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
+#[repr(C)]
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs]: We don't currently mark `usize` as zero-copy in wincode, so this won't have any effect on struct-level zero-copy.

We'd need to add some machinery that checks `target_pointer_width = 64`, integer config is zero-copy, integer endian config matches platform endian config

File: ledger/src/blockstore_meta.rs [hunk 13]
```diff
@@ -350,7 +381,8 @@ impl FrozenHashVersioned {
     }
 }
 
-#[derive(Deserialize, Serialize, Debug, PartialEq, Eq)]
+#[repr(C)]
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs]: `bool` is not zero-copy, so this wont have any effect on struct-level zero-copy

File: ledger/src/blockstore_meta.rs [hunk 14]
```diff
@@ -679,15 +711,16 @@ impl DuplicateSlotProof {
     }
 }
 
-#[derive(Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
+#[repr(transparent)]
```
Comment (cpubot) [inline ledger/src/blockstore_meta.rs]: `bool` is not zero-copy, so this wont have any effect on struct-level zero-copy

**Comment by kskalski** [comment]
> We do actually have `assert_zero_copy` https://docs.rs/wincode/latest/wincode/#top-level

Yap, its usefulness in practice is a bit limited:
* many structs use integers wider than 1 byte, so their zero-copy depends on the platform (possibly we could extend `assert_zero_copy` with endianess gate
* what we need is an eager detection that all fields meet zero-copy eligibility (for which endiness again..?) and lack of `repr` is the only missing part

But yes, given its existence, I think we can do better.

File: ledger/src/blockstore_meta.rs [hunk 15]
```diff
@@ -101,14 +102,14 @@ impl FromIterator<u32> for CompletedDataIndexes {
     }
 }
 
-// Manually implement Debut to display indices indices instead of raw u8's
+// Manually implement Debug to display indices instead of raw u8's
```
Comment (steviez) [inline ledger/src/blockstore_meta.rs:105]: Thank you 😅

**Review by cpubot — COMMENTED** [review]
LGTM

**Review by steviez — COMMENTED** [review]
Everything here looks in order. My only concern is some hypothetical edge case that could break compatibility.

Have you tried running a validator (with RPC metadata flag) and done upgrade/downgrade & ensure no issues ? The compat stuff doesn't show up in unit tests really since unit test don't persist state from "old format" to run with "new format"

**Comment by kskalski** [comment]
> Have you tried running a validator (with RPC metadata flag) and done upgrade/downgrade & ensure no issues ? The compat stuff doesn't show up in unit tests really since unit test don't persist 

Tested validator with
```
ENABLE_EXTENDED_TX_METADATA_STORAGE=true
ENABLE_RPC_TX_HISTORY=true
```
(does `FULL_RPC_API=false` matter here?)

restarting with and without PR, as well as running some `ledger-tool blockstore` commands in between.
No issues spotted.

**Comment by steviez** [comment]
> restarting with and without PR

Just making sure, you retained the blockstore between restarts right ?

> restarting with and without PR, as well as running some ledger-tool blockstore commands in between.

Can you run `agave-ledger-tool slot X -vv` for some blocks ? Namely, use pre-PR-ledger-tool to read blocks written by PR-validator. The `-vv` will ensure that the block metadata is all read out

**Comment by kskalski** [comment]
> Just making sure, you retained the blockstore between restarts right ?

fast-boot style restarts, all state is reused

> 
> Can you run `agave-ledger-tool slot X -vv` for some blocks ? Namely, use pre-PR-ledger-tool to read blocks written by PR-validator. The `-vv` will ensure that the block metadata is all read out

Yes, confirmed old and new ledger tool can read `slot -vv` for blocks produced by PR validator.

However there was a compilation error due to some test code from https://github.com/anza-xyz/agave/pull/11716, so I had to rebase + fix that.

**Review by steviez — COMMENTED** [review]
Cool, I think I saw @cpubot already approve this as well so LGTM

**Review by cpubot — APPROVED** [review]
